compiler/rustc_hir_typeck/src/cast.rs

```diff
@@ -803,40 +804,131 @@ impl<'a, 'tcx> CastCheck<'tcx> {
+
+                        // Check that the traits are actually the same
+                        // (this is required as the `Unsize` check below would allow upcasting, etc)
+                        // N.B.: this is only correct as long as we don't support `trait A<T>: A<()>`.
+                        //
+                        // Note that trait upcasting goes through a different mechanism (`coerce_unsized`)
+                        // and is unaffected by this check.
+                        if src_principal.def_id() != dst_principal.def_id() {
+                            return Err(CastError::DifferingKinds);
+                        }
 
-        // thin -> fat? report invalid cast (don't complain about vtable kinds)
-        if expr_kind == PointerKind::Thin {
-            return Err(CastError::SizedUnsizedCast);
-        }
+                        // We need to reconstruct trait object types.
+                        // `m_src` and `m_dst` won't work for us here because they will potentially
+                        // contain wrappers, which we do not care about.
+                        //
+                        // e.g. we want to allow `dyn T -> (dyn T,)`, etc.
+                        //
+                        // We also need to skip auto traits to emit an FCW and not an error.
+                        let src_obj = tcx.mk_ty_from_kind(ty::Dynamic(
+                            tcx.mk_poly_existential_predicates(
+                                &src_tty.without_auto_traits().collect::<Vec<_>>(),
+                            ),
+                            tcx.lifetimes.re_erased,
+                            ty::Dyn,
+                        ));
+                        let dst_obj = tcx.mk_ty_from_kind(ty::Dynamic(
+                            tcx.mk_poly_existential_predicates(
+                                &dst_tty.without_auto_traits().collect::<Vec<_>>(),
+                            ),
+                            tcx.lifetimes.re_erased,
+                            ty::Dyn,
+                        ));
 
-        // vtable kinds must match
-        if fcx.tcx.erase_regions(cast_kind) == fcx.tcx.erase_regions(expr_kind) {
-            Ok(CastKind::PtrPtrCast)
-        } else {
-            Err(CastError::DifferingKinds)
+                        // `dyn Src: Unsize<dyn Dst>`, this checks for matching generics
+                        let cause = fcx.misc(self.span);
+                        let obligation = Obligation::new(
+                            tcx,
+                            cause,
+                            fcx.param_env,
+                            ty::TraitRef::new(
+                                tcx,
+                                tcx.require_lang_item(LangItem::Unsize, Some(self.span)),
+                                [src_obj, dst_obj],
+                            ),
+                        );
+
+                        fcx.register_predicate(obligation);
```

Comment:
I still don't understand why we need to check for `Unsize` here rather than just type equality:

```suggestion
                        fcx.demand_eqtype(self.span, src_obj, dst_obj);
```

Are there any cases where these two are different? Is there a reason to support code like this (which is currently suggested by the compiler)?

```rust
#![feature(unsize)]
use core::marker::Unsize;

trait Trait<T> {}

fn cast<T, U>(ptr: *const dyn Trait<T>) -> *const dyn Trait<U>
where
    dyn Trait<T>: Unsize<dyn Trait<U>>,
{
    ptr as _
}
```

Reply:
So, the reason `Unsize` was used in the first place, is to check that auto traits in dst are a subset of auto traits in src. Now we have a different mechanism for that anyway, so this could probably be a type equality instead. Will look into it.